lambdas/preview/index.py

```diff
@@ -141,35 +147,95 @@ def extract_vcf(file_):
     (in that order, up to MAX_LINES)
-    """
 
+    Args:
+        file_ - file-like object
+    
+    Returns:
+        dict
+    """
     meta = []
     header = []
     data = []
+    size = 0
+    for line in file_:
+        line = _truncate(line.rstrip(), MAX_BYTES - size)
+        if line.startswith(b'##'):
+            meta.append(line)
+        elif line.startswith(b'#'):
+            header.append(line)
+        elif line:
+            data.append(line)
+        size += len(line)
+        total_lines = len(meta) + len(header) + len(data)
+        if total_lines >= MAX_LINES or size >= MAX_BYTES:
+            break
+
+    info = {
+        'data': {
+            # ignore b/c we might snap a multi-byte unicode character in two
+            'meta': [s.decode('utf-8', 'ignore') for s in meta],
+            'header': [s.decode('utf-8', 'ignore') for s in header],
+            'data': [s.decode('utf-8', 'ignore') for s in data]
+        }
+    }
 
-    with open(file_.name, 'rb') as vcf:
-        for index, line in enumerate(vcf, start=1):
-            # don't escape quotes
-            line = line.strip().decode('unicode_escape')
-            formatted = f'<p>{line}</p>'
-            if line.startswith('##'):
-                meta.append(formatted)
-            elif line.startswith('#'):
-                header.append(formatted)
-            else:
-                data.append(formatted)
+    return '', info
 
-            if index >= MAX_LINES:
-                break
+def extract_txt(file_):
+    """
+    Display first and last few lines of a potentially large file.
+    Because we process files in binary mode, we risk breaking some unicode
+    characters mid-word. Does NOT preview empty lines.
 
-    # pylint: disable=bad-continuation
-    html = ("<div>"
-                "<div data-type='meta'>{}</div>"
-                "<div data-type='header'>{}</div>"
-                "<div data-type='data'>{}</div>"
-            "</div>").format(
-                ''.join(meta) + '<br />',
-                ''.join(header) + '<br />',
-                ''.join(data))
+    Args:
+        file_ - file-like object
+    Returns:
+        dict - head and tail. tail may be empty. returns at most MAX_LINES
+        lines that occupy a total of MAX_BYTES bytes.
+    """
+    size = 0
+    head = []
+    tail = []
+    for index, line in enumerate(file_, start=1):
+        line = _truncate(line.rstrip(), MAX_BYTES - size)
+        size += len(line)
+        if line and len(head) < MAX_LINES:
+            head.append(line)
+        # if the file is longer than MAX_LINES, we need a tail
+        if index > MAX_LINES:
+            headmax = MAX_LINES//2
+            tailmax = MAX_LINES - headmax
+            # cut the head back to make room for the tail
+            head = head[:headmax]
+            # correct the size for the lines we just dropped
+            size = sum([len(s) for s in head])
+            file_size = os.stat(file_.name).st_size
```

Comment:
`os.fstat(file_.fileno())` or `file_.seek(0, os.SEEK_END)` would be better. In general, a temporary file might not even have a name. (It does in this particular case cause the Parquet parser needs it, and we're using `NamedTemporaryFile`, but it's good to not rely on that.)